packages/template-retail-react-app/app/contexts/index.js

```diff
@@ -40,65 +40,53 @@ export const CategoriesProvider = ({treeRoot = {}, children, locale}) => {
     const [root, setRoot] = useState({
-        // map over the server-provided cat
+        // Clone the server-provided category
         ...treeRoot[DEFAULT_ROOT_CATEGORY],
         [itemsKey]: treeRoot[DEFAULT_ROOT_CATEGORY]?.[itemsKey]?.map((item) => ({
             ...item
         }))
     })
 
     const fetchCategoryNode = async (id, levels = 1) => {
+        const storageKey = `${LOCAL_STORAGE_PREFIX}${id}-${locale}`
+        const storageItem = JSON.parse(window.localStorage.getItem(storageKey))
+
         // return early if there's one that is less than stale time
-        const storageItem = JSON.parse(
-            window.localStorage.getItem(`${LOCAL_STORAGE_PREFIX}${id}-${locale}`)
-        )
-        if (storageItem || Date.now() < storageItem?.fetchTime + CAT_MENU_STALE_TIME) {
+        if (storageItem && Date.now() < storageItem.fetchTime + CAT_MENU_STALE_TIME) {
             return storageItem
         }
+        // get and store fresh data for faster access / reduced server load
+        // TODO: Convert to useCategory hook when hooks are ready
         const res = await api.shopperProducts.getCategory({
             parameters: {
                 id,
                 levels
             }
         })
+        res.loaded = true
+        window.localStorage.setItem(
+            storageKey,
+            JSON.stringify({
+                ...res,
+                fetchTime: Date.now()
+            })
+        )
         return res
     }
 
-    useEffect(() => {
+    useEffect(async () => {
         // Server side, we only fetch level 0 categories, for performance, here
         // we request the remaining two levels of category depth
-        Promise.all(
-            root?.[itemsKey]?.map(async (cat) => {
-                // check localstorage first for this data to help remediate O(n) server
-                // load burden where n = top level categories
-                let res
-                try {
-                    res = await fetchCategoryNode(cat?.id, 2)
-                    // store fetched data in local storage for faster access / reduced server load
-                    res.loaded = true
-                    window?.localStorage?.setItem(
-                        `${LOCAL_STORAGE_PREFIX}${cat?.id}-${locale}`,
-                        JSON.stringify({
-                            ...res,
-                            fetchTime: Date.now()
-                        })
-                    )
-                    return res
-                } catch (error) {
-                    return error
-                }
-            })
-        )
-            .then((data) => {
-                const newTree = {
-                    ...root,
-                    [itemsKey]: data
-                }
-                setRoot(newTree)
-            })
-            .catch((err) => {
-                throw err
+        const promises = root?.[itemsKey]?.map(async (cat) => await fetchCategoryNode(cat?.id, 2))
+        if (promises) {
```

Comment:
Does the data from the API potentially lack `[itemsKey]`? What would be the consequence of having it default to an empty array if it is missing (way up in `useState`, not here)? We'd have to change relevant checks from `root?.[itemsKey]` to `root[itemsKey].length`, but how many of those are there? Are there other consequences that I'm not thinking of?

Reply:
`itemsKey` is a default constant set in the app, and the API will always respond with `categories` (which is the hardcoded `itemsKey`